os/arch/arm/src/amebasmart/amebasmart_idle.c

```diff
@@ -23,14 +23,160 @@
  ****************************************************************************/
 
 #include <tinyara/config.h>
-
 #include <tinyara/arch.h>
-#include "up_internal.h"
+#include <tinyara/irq.h>
+
+#ifdef CONFIG_PM
+#include <tinyara/pm/pm.h>
+#include "amebasmart_config.h"
+#include "arch_timer.h"
+#include "ameba_soc.h"
+#include "osdep_service.h"
+#endif
+
+#ifdef CONFIG_SMP
+#include "gic.h"
+#endif
 
 /****************************************************************************
- * Public Functions
+ * Name: up_idlepm
+ *
+ * Description:
+ *   Perform IDLE state power management.
+ *
  ****************************************************************************/
 
+#ifdef CONFIG_PM
+static enum pm_state_e oldstate = PM_NORMAL;
+static void up_idlepm(void)
+{
+	uint32_t xModifiableIdleTime = 0;
+	enum pm_state_e newstate;
+	irqstate_t flags;
+	int ret;
+
+	/* Decide, which power saving level can be obtained */
+	newstate = pm_checkstate(PM_IDLE_DOMAIN);
+
+	/* Check for state changes */
+	if (newstate != oldstate)
+	{
+		//TODO: Critical section code needed for SMP case?
+		//Any additional implications of putting a core in critical section while trying to sleep?
+		/* Perform board-specific, state-dependent logic here */
+	  	pmvdbg("newstate= %d oldstate=%d\n", newstate, oldstate);
+
+		/* Then force the global state change */
+		ret = pm_changestate(PM_IDLE_DOMAIN, newstate);
+		if (ret < 0) {
+			/* The new state change failed, revert to the preceding state */
+			pmdbg("State change failed! Current state = %d, newstate = %d\n", oldstate, newstate);
+			newstate = oldstate;
+		} else {
+			/* Save the new state */
+			oldstate = newstate;
+		}
+		/* MCU-specific power management logic */
+		switch (newstate) {
+			case PM_NORMAL:
+				break;
+			case PM_IDLE:
+				break;
+			case PM_STANDBY:
+				break;
+			case PM_SLEEP:
+				ap_timer_helper();
+				if (up_cpu_index() == 0) {
+					/* mask sys tick interrupt*/
+					arm_arch_timer_int_mask(1);
+					up_timer_disable();
+					flags = irqsave();
+					if (tizenrt_ready_to_sleep()) {
+// Consider for dual core condition
```

Comment:
This was another comment that I added as part of the skeleton, this should be resolved and removed from here.